plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/TypeJudgment.java

```diff
@@ -565,6 +573,39 @@ public TypeRef caseArrayElement(ArrayElement e) {
 		public TypeRef caseIdentifierRef(IdentifierRef idref) {
 			TypeRef T = ts.type(G, idref.getId());
 
+			final ASTFlowInfo flowInfo = ((N4JSResource) idref.eResource()).getASTMetaInfoCache().getFlowInfo();
+			final Collection<InstanceofGuard> alwaysHoldingTypes = flowInfo.instanceofGuardAnalyser
+					.getAlwaysHoldingTypes(idref);
+
+			if (!alwaysHoldingTypes.isEmpty()) {
+				final Collection<TypeRef> intersectionTypes = new LinkedList<>();
+				for (InstanceofGuard ioGuard : alwaysHoldingTypes) {
+					if (ioGuard.symbolCFE instanceof IdentifierRef) {
+						final IdentifiableElement guardedElement = ((IdentifierRef) ioGuard.symbolCFE).getId();
+						if (guardedElement != null && idref.getId() == guardedElement) {
+							final Expression typeIdentifier = ioGuard.typeIdentifier;
+							TypeRef instanceofType = ts.type(G, typeIdentifier);
+
+							if (tsh.isClassConstructorFunction(G, instanceofType)
+									&& instanceofType instanceof TypeTypeRef) {
+
+								final TypeTypeRef ttRef = (TypeTypeRef) instanceofType;
+								final TypeArgument typeArg = ttRef.getTypeArg();
+								if (typeArg instanceof TypeRef) {
+									instanceofType = (TypeRef) typeArg;
+								}
+							}
+
+							TypeUtils.sanitizeRawTypeRef(instanceofType);
+							intersectionTypes.add(instanceofType);
+						}
+					}
+				}
+				if (!intersectionTypes.isEmpty()) {
+					T = tsh.createIntersectionType(G, intersectionTypes.toArray(new TypeRef[intersectionTypes.size()]));
```

Comment:
Another example using generics instead of interfaces:
```javascript
class G<T> {
	getIt(): T { return null; }
}
class H<T> extends G<T> {}

function foo(g: G<string>) {
	
	if(g instanceof H) {
		let s: string;
		// XPECT noerrors --> "? is not a subtype of string." at "g.getIt()"
		s = g.getIt();
	}
}
```

Reply:
Thanks for the examples! I added another xpect test file "TypeGuards.n4js.xt" to test that. Also, I adjusted the type judgement to deal with these situations.